---
title: "Class18: Pertussis and the CMI-PB project"
author: "Alice Lai (PID:A16799081)"
format: pdf
---

## 1. Investigating pertussis cases by year
```{r}
#install.packages("datapasta")

library(datapasta)
library(ggplot2)

cdc <- data.frame(
  Year = c(1922, 1923, 1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 
           1934, 1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 
           1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 
           1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 
           1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 
           1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 
           1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 
           2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 
           2018, 2019, 2020, 2021),
  Cases = c(107473, 164191, 165418, 152003, 202210, 181411, 161799, 197371, 166914, 
            172559, 215343, 179135, 265269, 180518, 147237, 214652, 227319, 103188, 
            183866, 222202, 191383, 191890, 109873, 133792, 109860, 156517, 74715, 
            69479, 120718, 68687, 45030, 37129, 60886, 62786, 31732, 28295, 32148, 
            40005, 14809, 11468, 17749, 17135, 13005, 6799, 7717, 9718, 4810, 3285, 
            4249, 3036, 3287, 1759, 2402, 1738, 1010, 2177, 2063, 1623, 1730, 1248, 
            1895, 2463, 2276, 3589, 4195, 2823, 3450, 4157, 4570, 2719, 4083, 6586, 
            4617, 5137, 7796, 6564, 7405, 7298, 7867, 7580, 9771, 11647, 25827, 25616, 
            15632, 10454, 13278, 16858, 27550, 18719, 48277, 28639, 32971, 20762, 17972, 
            18975, 15609, 18617, 6124, 2116)
)


ggplot(cdc, aes(x = Year, y = Cases)) +
  geom_point() +
  geom_line() +
  labs(title = "Pertussis Cases by Year (1922-2021)",
       x = "Year",
       y = "Number of cases")

```


## 2. A tale of two vaccines (wP & aP)

```{r}
ggplot(cdc, aes(x = Year, y = Cases)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 1946, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 1996, linetype = "dashed", color = "red") +
  geom_text(aes(x = 1946, y = max(Cases), label = "wP"), color = "blue", hjust = -0.5) +
  geom_text(aes(x = 1996, y = max(Cases), label = "aP"), color = "red", hjust = -0.5) +
  labs(title = "Pertussis Cases by Year (1922-2021)",
       x = "Year",
       y = "Number of cases")
```

**Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?**

According to CDC data, pertussis cases are on the rise again. For instance, in 2012, the CDC reported 48,277 cases of pertussis in the U.S., the highest number since 1955, when 62,786 cases were reported. Experts in the field suggest several possible reasons for this resurgence, including: (1) the increased sensitivity of PCR-based testing, (2) vaccine hesitancy, (3) bacterial evolution allowing escape from vaccine-induced immunity, and (4) diminishing immunity in adolescents who were initially vaccinated with the newer aP vaccine compared to those who received the older wP vaccine.


## 3. Exploring CMI-PB data

```{r}
# Allows us to read, write and process JSON data
library(jsonlite)

subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 

head(subject, 3)
```

**Q4. How many aP and wP infancy vaccinated subjects are in the dataset?**

```{r}
table(subject$infancy_vac)
```


**Q5. How many Male and Female subjects/patients are in the dataset?**

```{r}
table(subject$biological_sex)
```


**Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?**

```{r}
table(subject$race, subject$biological_sex)
```

```{r}
#install.packages("lubridate")
library(lubridate)

today()
today() - ymd("2000-01-01")
time_length( today() - ymd("2000-01-01"),  "years")
```


**Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?**

```{r}
# Use todays date to calculate age in days
subject$age <- today() - ymd(subject$year_of_birth)
```

```{r}
library(dplyr)
# aP
ap <- subject %>% filter(infancy_vac == "aP")
round( summary( time_length( ap$age, "years" ) ) )
```
```{r}
# wP
wp <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wp$age, "years" ) ) )
```

The average age for wP individuals is 37 years, and for aP is 27 years. 


**Q8. Determine the age of all individuals at time of boost?**

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```


**Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?**

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```

```{r}
x <- t.test(time_length( wp$age, "years" ),
       time_length( ap$age, "years" ))

x$p.value
```

From the plot and the p-value, it is obvious that the two groups are significantly different. 


### Joining multiple tables

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/plasma_ab_titer", simplifyVector = TRUE) 
```

**Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:**

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

**Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.**

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

**Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?**

```{r}
table(abdata$isotype)
```

**Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?**
```{r}
table(abdata$dataset)
```
The most recent one has least number of columns. 


## 4. Examine IgG Ab titer levels

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

**Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:**

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) + 
  geom_boxplot() +
  xlim(0, 75) +
  facet_wrap(vars(visit), nrow = 2)  
```

**Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?**

TT and DT show high and consistent IgG antibody titers, likely due to regular booster vaccinations.
PRN and FHA show higher variability, possibly due to differences in vaccine formulations and individual responses.
Fim2/3 show lower and less variable titers, suggesting they might be less immunogenic.


```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```

```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```


**Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. **

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = F) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(igg, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = F) +
  facet_wrap(vars(visit)) +
  theme_bw()
```


**Q16. What do you notice about these two antigens time courses and the PT data in particular?**

PT level increased over time and is significantly higher than OVA levels. 


**Q17. Do you see any clear difference in aP vs. wP responses?**

aP appeared to respond slower than wP did. 


```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```


**Q18. Does this trend look similar for the 2020 dataset?**

Yes, this trend look similar for the 2020 dataset.

```{r}
abdata.20 <- abdata %>% filter(dataset == "2020_dataset")

abdata.20 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

## 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

**Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).**

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

**Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?**

The expression of this gene is at its max level at visit 4. 



```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```



## 6. Working with larger datasets [OPTIONAL]